cloudstate/cloudstate.go

```
@@ -0,0 +1,249 @@
+//
+// Copyright 2019 Lightbend Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cloudstate
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/cloudstateio/go-support/cloudstate/protocol"
+	"github.com/golang/protobuf/descriptor"
+	"github.com/golang/protobuf/proto"
+	filedescr "github.com/golang/protobuf/protoc-gen-go/descriptor"
+	"github.com/golang/protobuf/ptypes/empty"
+	"google.golang.org/grpc"
+	"log"
+	"net"
+	"os"
+	"runtime"
+)
+
+const (
+	SupportLibraryVersion = "0.1.0"
+	SupportLibraryName    = "cloudstate-go-support"
+)
+
+// CloudState is an instance of a CloudState User Function
+type CloudState struct {
+	server                   *grpc.Server
+	entityDiscoveryResponder *EntityDiscoveryResponder
+	eventSourcedHandler      *EventSourcedHandler
+}
+
+// NewCloudState returns a new CloudState instance.
+func NewCloudState(options *Options) *CloudState {
```

Can you rename this to `New`? Per [effective go](https://golang.org/doc/effective_go.html#package-names), it's redundant to do `packagename.NewPackageName` - prefer `packagename.New` instead.

---

Also consider changing this to take in a value rather than a pointer, to enforce users to provide the ServiceName/ServiceVersion rather than a nil.

---

Should we also return an error here? Could we ever in the future do something inside New that may fail?

---

- [ ] rename NewCloudState to New
- [ ] return an error for NewCloudState